project hackyourweather/hackyourweather/src/Components/SearchComponent/index.js

```diff
@@ -1,52 +1,50 @@
-import React, { useState, useEffect } from "react";
-import './styles.css'
-import CityBox from './../CityBox'
-
-
-
-const APIKEY = '4bde5fff659a96c59438421c777e914e'
-
+import React, { useState } from 'react';
+import './styles.css';
+import CityBox from './../CityBox';
 
+const APIKEY = '4bde5fff659a96c59438421c777e914e';
 
 function SearchComponent() {
+	let [ cityName, setCityName ] = useState('');
+	let [ cityData, setCityData ] = useState([]);
 
-    let [cityName, setCityName] = useState('')
-    let [cityData, setCityData] = useState()
-
-
+	const apiUrl = `https://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${APIKEY}`;
 
+	function submitHandler(e) {  
+		e.preventDefault();
 
+        fetch(apiUrl)
+            .then((response) => response.json())
+            .then((data) => {
+                if (data.cod !== '404') {
 
-    const apiUrl = `http://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${APIKEY}`
-
-    async function submitHandler(e) {
-        e.preventDefault()
-
-        await fetch(apiUrl)
-            .then(response => response.json())
-            .then(data => {
-                setCityData(data)
-                console.log('the setted data is ', data)
+                    setCityData([...cityData, data]);
+                }
+                else if (data.cod ==='404') { alert("City can't be found, please check your spelling") }
 
             })
 
-    }
+        setCityName('')
 
-    useEffect(() => {
+    }
 
-    }, [cityData])
+    function removeCity(props) {
 
+        if (cityData.length > 0) {
+            setCityData(cityData.filter(city => city.id !== props.id))
+        }
+    }
 
 
-    return (
-        <div >
-            <form onSubmit={submitHandler} className="searchBar">
-                <input type='text' required placeholder='Search City' onChange={(e) => setCityName(e.target.value)} />
-                <input type='submit' value='Search' />
-            </form>
-            {cityData && <CityBox data={cityData} />}
-        </div>
-    )
+	return (
+		<div>
+			<form onSubmit={submitHandler} className="searchBar">
+                <input type="text" value={cityName} required placeholder="Search City" onChange={(e) => setCityName(e.target.value)} />
+				<input type="submit" value="Search" />
+			</form>
+            {cityData && cityData.map((city, index) => { return <CityBox key={index} id={city.id}  data={city} removeCity={removeCity}/>})}
```

Comment:
cityData will never be null or undefined. So `cityData.map` will never throw an error. If the array is empty it won't render anything

Reply:
Don't pass props that you later don't use. eg `id`

Reply:
You can create an arrow function here to delete the city. eg
```suggestion
            {cityData.map((city, index) => (
                <CityBox key={city.id} data={city} removeCity={() => removeCity(city.id)}/>
            )}
```
And it the CityBox component you can just say
`<span className="removeButton" onClick={removeCity} >X</span>`

Reply:
also when creating a key in a map we first try to find a unique value in our data before reaching the map index. 
Here you have the city.id that is unique. The index of the array should be our last option

Reply:
yes but i noticed than when i was using the city.id i had this problem : lets say i put 4 cities :
-athens
-london
-berlin
-athens

If i click on the last ' athens' it will automatically delete the first one also, since they have the same id in their props. so could i use index instead?

Reply:
First of all you delete the city based on the id of the city. So if someone clicks on `athens` both instances of athens will be deleted. In order to solve this you can delete using the index in the array.
The other thing you need to decide is whether or not you will allow a city to appear twice in your list. 
If you decide that it can appear multiple times you will need to use the index as the key and remove the entry be the index as well.
if you decide that there can be only one instance of a city in this list you can use the city.id as key and remove using the city.id as well. In this case you need to pay special attention when inserting a city into the array because you would have to remove any previous instance of the city before adding it.
As a last thought do you add the city as last or first on the list? Which is preferable for the user?